File: frontend/src/types/sale.ts
export interface SaleItem {
  id: number
  product_id: number
  quantity: number
  unit_price: number
  line_total: number
  product: { id: number; name: string; sku: string; selling_price?: number; category: { id: number; name: string } }
}

export interface Sale {
  id: number
  invoice_number: string
  customer_name?: string
  customer_phone?: string
  subtotal: number
  tax_amount: number
  discount: number
  total_amount: number
  payment_method: string
  status: string
  notes?: string
  sale_date: string
  sold_by: number
  created_at: string
  items: SaleItem[]
}

export interface SalesSummary {
  today_sales: number
  monthly_sales: number
  today_count: number
  monthly_count: number
}

export interface Paged<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

export interface CartLine {
  product_id: number
  name: string
  sku: string
  quantity: number
  unit_price: number
}

File: backend/app/schemas/sale.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.catalog import ProductResponse


class SaleItemCreate(BaseModel):
    product_id: int
    quantity: int = Field(gt=0)
    unit_price: float | None = Field(default=None, ge=0, description="Defaults to product selling_price")


class SaleCreate(BaseModel):
    customer_name: str | None = None
    customer_phone: str | None = None
    payment_method: str = Field(default="cash", description="cash|card|upi|credit")
    tax_percent: float = Field(default=0, ge=0, le=100)
    discount: float = Field(default=0, ge=0)
    notes: str | None = None
    items: list[SaleItemCreate] = Field(min_length=1)


class SaleItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    quantity: int
    unit_price: float
    line_total: float
    product: ProductResponse


class SaleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    invoice_number: str
    customer_name: str | None = None
    customer_phone: str | None = None
    subtotal: float
    tax_amount: float
    discount: float
    total_amount: float
    payment_method: str
    status: str
    notes: str | None = None
    sale_date: date
    sold_by: int
    created_at: datetime
    items: list[SaleItemResponse]


class SalesSummary(BaseModel):
    today_sales: float
    monthly_sales: float
    today_count: int
    monthly_count: int
